nbviewer/app.py

```diff
@@ -80,196 +81,200 @@ def nrfoot():
 this_dir, this_filename = os.path.split(__file__)
 FRONTPAGE_JSON = os.path.join(this_dir, "frontpage.json")
 
+class NBViewer(Application):
 
-def make_app():
-    # NBConvert config
-    config = Config()
-    config.NbconvertApp.fileext = 'html'
-    config.CSSHTMLHeaderTransformer.enabled = False
-    # don't strip the files prefix - we use it for redirects
-    # config.Exporter.filters = {'strip_files_prefix': lambda s: s}
+    config_file = Unicode('nbviewer_config.py', help="The config file to load").tag(config=True)
```

Comment:
Could you please add a short note in the README about where `nbviewer_config.py` must reside and it's format, maybe via an example of what one might want to customize using the config file?

I don't think this refactor makes NBViewer a first class JupyterApplication supporting the common config outlined here https://jupyter.readthedocs.io/en/latest/projects/config.html, but please correct me if I'm wrong.

Reply:
It depends on what one means by a first class JupyterApplication. You are correct that right now the only configurable attribute is the name of the config file. Right now it works the same way JupyterHub's config does -- as long as the file is in the directory from which the nbviewer program is executed, it will pick up the config file. 

I wanted to keep the changes for now as small as possible so that: (1) the chances something which is breaking goes unnoticed are minimized, and (2) the PR is as easy to review as possible. I.e. you are correct, by itself it does mostly nothing, but that's the point. It doing mostly nothing means it's easier to scrutinize, but at the same time still manages to step towards the direction of a first class JupyterApplication.

In order to make NBViewer meaningfully configurable it would be necessary, among other things, to refactor the class to have attributes which are traitlet configurables. In [the next branch of my fork, step2](https://github.com/krinsman/nbviewer/blob/step2/nbviewer/app.py) (a modified version of which I would make as a new PR if this PR was accepted), the `make_app` is refactored so that most of the computed quantities are now attributes (properties) of the NBViewer class. A further PR would then have to identify which of those attributes/properties to make instances of traitlets classes with the `.tag(config=True)` tag (in order to make them configurable from the config file). 

Also another way in which this does not yet make NBViewer a first class JupyterApplication is that it does not yet allow the configurables to be changed by adding command line flags (maybe this is in particular what you had in mind?), so e.g. `python -m nbviewer --c.NBViewer.config_file='better_name.py'` wouldn't work. In JupyterHub, that happens [here](https://github.com/jupyterhub/jupyterhub/blob/3089d441b4d499c2b5662acf8dd8c95837f3490f/jupyterhub/app.py#L183), and is mostly implemented in the traitlets source code [here](https://github.com/ipython/traitlets/blob/master/traitlets/config/application.py#L666). So I do have an idea of how to refactor NBViewer further such that configurable options can be parsed via the command line in that manner, but that for one thing would involve mucking with the 30+ lines of code of `define` function calls currently in `app.py`, which would make the PR hard to read, and which would increase the possibility of accidentally introducing breaking changes.

Regarding what one might want to customize, again with this little PR there really isn't anything to customize yet, you are correct, since there aren't any configurable attributes (besides `config_file` itself). But in later steps (e.g. see [the `latest` branch of my fork](https://github.com/krinsman/nbviewer/blob/latest/nbviewer/app.py#L90), which if this PR is step1, would be around step6 or step7, I think) it's possible to make the handlers for all of the different providers customizable. So, for example, if one wants to subclass the existing handlers, customize them to add extra logic (e.g. for cloning notebooks into a home directory), and then have NBViewer use those custom handlers instead, one can do that. I have the code required for that implementation [here](https://github.com/krinsman/clonenotebooks) (as well as in the `latest` branch of my NBViewer fork); the setup is complicated so I created a [complete Docker image](https://github.com/krinsman/jupyterhub-deploy/tree/master/nbviewer/cloneviewer) which should make it possible to demo locally the type of future that this PR is a step towards, just by running `build.sh`.

Sorry, I've been working on this for about 1.5 months now, it's hard for me to talk about it succinctly since I've thought about it way too much.

Reply:
What you added to the README are the details I was hoping to surface in this PR. Thanks!